Experimentally emit diagnostics from the new Swift parser

DougGregor opened the pull request:
Introduce the experimental feature `ParserDiagnostics`, which emits
diagnostics from the new Swift parser *first* for a source file. If
that produces any errors, we suppress any diagnostics emitted from the
C++ parser.

While here, move diagnostic validation checking into ASTGen, so we
can drop our dependency on SwiftCompilerSupport... which will go away.

DougGregor committed:
[ASTGen] Handle new parser validation in ASTGen.

Replace the use of the "consistency check" vended by swift-syntax with
an ASTGen-implemented operation that emits diagnostics from the new parser
via the normal diagnostic engine. This eliminates our last dependency
on SwiftCompilerSupport, so stop linking it.

DougGregor committed:
Experimentally emit diagnostics from the new Swift parser

Introduce the experimental feature `ParserDiagnostics`, which emits
diagnostics from the new Swift parser *first* for a source file. If
that produces any errors, we suppress any diagnostics emitted from the
C++ parser.

DougGregor committed:
[ASTGen] Make sure we handle Fix-Its without a matching note.

Any Fix-Its not associated with a note are automatically associated
with the main diagnostic.

DougGregor commented:
@swift-ci please smoke test

bnbarham commented on lib/ASTGen/Sources/ASTGen/Diagnostics.swift:
```diff
@@ -91,10 +91,18 @@ func emitDiagnostic(
   diagnostic: Diagnostic,
   messageSuffix: String? = nil
 ) {
+  // Determine the set of note IDs.
+  let knownNoteIDs: Set<MessageID> = .init(
+    diagnostic.notes.map { $0.noteMessage.fixItID }
+  )
```
The Swift parser models diagnostics a little differently. Each diagnostic has both notes + fix-it's. A note has no fix-it attached (I think `fixItID` is actually misnamed here, it's just the note ID). A fix-it has both a message (or "note" if you like :P) and a set of changes.

We can talk about whether that's how we want to model that or not later, but IMO the way these should be emitted right now is that each of the fix-it's just become a note with an attached fix-it.

DougGregor replied:
This doesn't match the model used by the compiler, and is somewhat lossy. But I'll embrace the model here (see my latest commit) and we can iterate on the swift-syntax diagnostic model more.

DougGregor committed:
Embrace the swift-syntax notion of Fix-Its as separate notes.

The swift-syntax diagnostic system always treats Fix-Its as separate
notes, which are never attached to the primary diagnostic. Embrace this
module in the mapping over to the existing C++ diagnostic engine.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test

ahoppen commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -2021,6 +2015,9 @@ ERROR(macro_expansion_decl_expected_macro_identifier,none,
 ERROR(parser_round_trip_error,none,
       "source file did not round-trip through the Swift Swift parser", ())
+ERROR(parser_new_parser_errors,none,
+      "Swift Swift parser generated errors for code that C++ parser accepted",
```
Similar to my comment on another PR, I think `Swift Swift parser` is confusing. What about `SwiftParser` or `SwiftParser implemented in Swift`?
